### PR TITLE
Room version abstractions

### PR DESCRIPTION
**Pull request by neilalexander**
This PR (still a work in progress) adds some support for differentiating between room versions and state resolution algorithms.

### CODE REVIEW

File: clientapi/routing/capabilities.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
+// Copyright 2017 Vector Creations Ltd
```
Comment (turt2live) [inline clientapi/routing/capabilities.go]: *ahem*

Comment (neilalexander) [reply clientapi/routing/capabilities.go]: Thanks for the catch - I'll fix 💃

**Comment by neilalexander** [comment]
Notes for review:

- This is one of two broken out PRs now. Sytest will possibly fail but the fixes to the whitelist/blacklist are in the next PR
- Most of the changes are in moving `roomserver/state/state.go` into `roomserver/state/v1/state.go` and changing some things to use function receivers so that the `db` reference is internal (GitHub makes it look like a huge change but in reality it's not much)

File: roomserver/query/query.go [hunk 2]
```diff
@@ -98,6 +99,11 @@ func (r *RoomserverQueryAPI) QueryLatestEventsAndState(
 	request *api.QueryLatestEventsAndStateRequest,
 	response *api.QueryLatestEventsAndStateResponse,
 ) error {
+	// TODO: get the correct room version
+	state, err := state.GetStateResolutionAlgorithm(state.StateResolutionAlgorithmV1, r.DB)
```
Comment (babolivier) [inline roomserver/query/query.go]: Not a huge fan of the fact that this variable is shadowing the `state` package, but I guess convenience makes it worthwhile here.

**Pull request merged**